### PR TITLE
Test base signature

### PR DESCRIPTION
**Pull request by shangrex**

### CODE REVIEW

File: test/lmp/dset/_ch_poem/test_signature.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+                annotation=Optional[str],
+                default=None,
+            ),
+        ],
+        return_annotation=Signature.empty,
+    )
+
+
+def test_inherent_method():
+    r'''Ensure inherent methods are same as baseclass.'''
```
Comment (shangrex) [inline test/lmp/dset/_ch_poem/test_signature.py]: Wrong comment

File: test/lmp/dset/conftest.py [hunk 2]
```diff
@@ -0,0 +1 @@
+r"""Setup fixture for testing :py:mod:`lmp.dset`."""
```
Comment (shangrex) [inline test/lmp/dset/conftest.py]: remove empty file.

File: test/lmp/dset/util/test_signature.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
+r"""Test :py:class:`lmp.dset.ChPoemDset` signature."""
+
+import inspect
+from inspect import Parameter, Signature
+from typing import (Optional)
+
+from lmp.dset.util import download, norm, trunc_to_max, pad_to_max
+
+
+def test_func():
```
Comment (shangrex) [inline test/lmp/dset/util/test_signature.py]: Do not use abbr. in test files.

File: test/lmp/dset/util/test_signature.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    r"""Ensure abstract func signature.
```
Comment (shangrex) [inline test/lmp/dset/util/test_signature.py]: Wrong comments.

File: test/lmp/infer/_base/conftest.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+from lmp.infer._base import BaseInfer
+from lmp.model._base import BaseModel
+from lmp.tknzr._base import BaseTknzr
+
+
+@pytest.fixture
+def subclss_infer_clss() -> Type[BaseInfer]:
+    r"""Simple ``BaseInfer`` subclass."""
+    class SubclssInfer(BaseInfer):
+        r"""Only implement ``gen`` ."""
```
Comment (shangrex) [inline test/lmp/infer/_base/conftest.py]: redundant whitespace

Comment (shangrex) [reply test/lmp/infer/_base/conftest.py]: there is no method  implemented, so no need to test.

File: test/lmp/infer/_base/test_signature.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    r"""Ensure abstract method's signature."""
+    assert 'gen' in BaseInfer.__abstractmethods__
+
+
+def test_instance_attribute(
+        max_seq_len: int,
+        subclss_infer: BaseInfer,
+):
+    r"""Ensure instance attributes' signature."""
+    assert subclss_infer.max_seq_len == max_seq_len
```
Comment (shangrex) [inline test/lmp/infer/_base/test_signature.py]: Directly create `BaseInfer` instance.

File: test/lmp/dset/_ch_poem/test_signature.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    assert ChPoemDset.url == ''.join([
+        'https://github.com/ProFatXuanAll',
+        '/demo-dataset/raw/main/ch-poem',
+    ])
+
+
+def test_instance_method():
+    r"""Ensure instance methods' signature."""
+    assert hasattr(ChPoemDset, '__init__')
+    assert inspect.signature(ChPoemDset.__init__) == Signature(
```
Comment (shangrex) [inline test/lmp/dset/_ch_poem/test_signature.py]: Ensure consistency on inherit methods

File: test/lmp/dset/_ch_poem/test_signature.py [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+        inspect.signature(BaseDset.__iter__)
+        == inspect.signature(ChPoemDset.__iter__)
+    )
+
+    assert (
+        inspect.signature(BaseDset.__len__)
+        == inspect.signature(ChPoemDset.__len__)
+    )
+
+    assert (
```
Comment (shangrex) [inline test/lmp/dset/_ch_poem/test_signature.py]: sort by alphabetic order

File: doc/source/contribute.rst [hunk 9]
```diff
@@ -57,11 +57,12 @@ Before You Commit
    .. code-block:: sh
 
         autopep8 -i -r -a -a -a lmp test
+        isort test
         flake8 lmp test
 
    Fix any error/warning messages showed on CLI.
    If you find some rule is not possible to be fix, please **open issue**.
-
+
```
Comment (ProFatXuanAll) [inline doc/source/contribute.rst:65]: remove redundant whitespaces

File: test/lmp/infer/_top_1/test_signature.py [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+    assert inspect.isclass(Top1Infer)
+    assert not inspect.isabstract(Top1Infer)
+    assert issubclass(Top1Infer, BaseInfer)
+
+
+def test_class_attribute():
+    r"""Ensure class attributes' signature."""
+    print(get_type_hints(Top1Infer))
+    assert get_type_hints(Top1Infer) == get_type_hints(BaseInfer)
+    assert Top1Infer.hard_max_seq_len == 512
```
Comment (ProFatXuanAll) [inline test/lmp/infer/_top_1/test_signature.py:21]: `assert Top1Infer.hard_max_seq_len == BaseInfer.hard_max_seq_len`

File: test/lmp/infer/_top_k/test_signature.py [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+    assert inspect.isclass(TopKInfer)
+    assert not inspect.isabstract(TopKInfer)
+    assert issubclass(TopKInfer, BaseInfer)
+
+
+def test_class_attribute():
+    r"""Ensure class attributes' signature."""
+    print(get_type_hints(TopKInfer))
+    assert get_type_hints(TopKInfer) == get_type_hints(BaseInfer)
+    assert TopKInfer.hard_max_seq_len == 512
```
Comment (ProFatXuanAll) [inline test/lmp/infer/_top_k/test_signature.py:22]: `assert TopKInfer.hard_max_seq_len == BaseInfer.hard_max_seq_len`

File: test/lmp/model/_base/conftest.py [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+import torch
+
+from lmp.model._base import BaseModel
+
+
+@pytest.fixture
+def subclss_model_clss() -> Type[BaseModel]:
+    r"""Simple ``BaseModel`` subclass."""
+    class SubclssModel(BaseModel):
+        r"""Only implement ``tknz`` and ``dtknz``."""
```
Comment (ProFatXuanAll) [inline test/lmp/model/_base/conftest.py:15]: Wrong comment

File: test/lmp/model/_base/conftest.py [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+
+        def forward(self, batch_prev_tkids: torch.Tensor) -> torch.Tensor:
+            return ''.join(batch_prev_tkids)
+
+        def loss_fn(
+                self,
+                batch_next_tkids: torch.Tensor,
+                batch_prev_tkids: torch.Tensor,) -> torch.Tensor:
```
Comment (ProFatXuanAll) [inline test/lmp/model/_base/conftest.py:23]: Linting issue

File: test/lmp/model/_base/conftest.py [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+
+
+@pytest.fixture
+def subclss_model_clss() -> Type[BaseModel]:
+    r"""Simple ``BaseModel`` subclass."""
+    class SubclssModel(BaseModel):
+        r"""Only implement ``tknz`` and ``dtknz``."""
+
+        def forward(self, batch_prev_tkids: torch.Tensor) -> torch.Tensor:
+            return ''.join(batch_prev_tkids)
```
Comment (ProFatXuanAll) [inline test/lmp/model/_base/conftest.py:18]: Return dummy tensor with shape `(B, S, V)`.
For example: `torch.zeros(batch_prev_tkids.size(0), batch_prev_tkids.size(1), 100)`

File: test/lmp/model/_base/conftest.py [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+
+        def loss_fn(
+                self,
+                batch_next_tkids: torch.Tensor,
+                batch_prev_tkids: torch.Tensor,) -> torch.Tensor:
+            return ''.join(batch_prev_tkids)
```
Comment (ProFatXuanAll) [inline test/lmp/model/_base/conftest.py:24]: Return dummy tensor with shape `(1)`.
For example: `torch.zeros(1)`

**Pull request closed without merging**